enhancements/network/user-space-pod-iface-and-api-library.md

```diff
@@ -0,0 +1,262 @@
+- SR-IOV Device Plugin
+- SR-IOV CNI
+- SR-IOV Network-Resource Injector
+- Multus
+- app-netutil
+
+#### Consumption of Data in the Pod
+
+Data passed into a pod can come from several places. Below are some examples:
+- For kernel interfaces, data can be applied to the interface via the kernel
+itself.
+- For SR-IOV VFs, the PCI Address list is passed into the pod via an
+environmental variable.
+- Using Kubernetes Downward API, pod annotation data is exposed within the pod.
+Data is written to the pod annotation and consumed by the pod.
+
+To ease the burden of the CNF developer, an API library is being developed that
+collects data from all the different sources and exposes the data through a set
+of APIs. The API is written in Golang and can be imported into other Golang
+programs. Since DPDK is the primary user space application being targeted, and
+DPDK is written in C, there is also a C API binding for the Golang library.
+
+The library is located in github at the following URL:
+- https://github.com/openshift/app-netutil/
+
+The library will not be released in OCP, but will be available for CNF
+developers to include in their projects. If the CNF developers choose not to
+include the library, the library will still serve as an example of where to pull
+the data from and how to consume the data.
+
+The following APIs will be exposed:
+- GetCPUInfo()
+  - Returns the set of CPUs assigned to the pod.
+- GetInterfaces()
+  - Returns the set of interfaces associated with the pod and a structure for
+  each interface containing the type of interface and the data specific to that
+  interface type.
+- GetHugepageInfo()
+  - Returns the amount of hugepage memory reserved for the pod and page size.
+
+### Implementation Details/Notes/Constraints [optional]
+
+### Risks and Mitigations
+
+## Design Details
+
+### Open Questions [optional]
+
+### Test Plan
+
+### Graduation Criteria
+
+The '*appNetUtil*' library was introduced as Tech Preview in OCP 4.5. The plan
+is to make the library GA in OCP 4.7. To our knowledge, the '*appNetUtil*'
+library is not currently being used. With the additions described in this
+enhancement, the library should be more useful to CNF developers. Also by moving
+to GA, the library attract more users and provide feedback on ways to improve
+the usefulness  of the library.
+
+##### Dev Preview -> Tech Preview
+
+In Tech Preview, the following APIs are defined in
+"*github.com/openshift/app-netutil/lib/v1alpha*":
+- GetCPUInfo()
+  - Returns the set of CPUs assigned to the pod.
+- GetInterfaces()
+  - Returns the set of interfaces associated with the pod and a structure for
+  each interface containing the type of interface and the data specific to that
+  interface type.
+
+##### Tech Preview -> GA 
+
+In GA, the following APIs are defined in
+"*github.com/openshift/app-netutil/lib/v1*":
+
+- GetCPUInfo()
+  - Returns the set of CPUs assigned to the pod.
+  - Same as Tech Preview.
+- GetInterfaces()
```

Comment:
Do we need to complete the additional device information proposal in 4.7 or it can come later?

Reply:
We don't need all the data, but the proposal in the NPWG adds the DeviceId to the network-status annotaion, which is the key to tie a given PCI Address (interface) with the network (how the interface should be used).

Reply:
ack, so we will need the proposal be implemented to get the pci-network mapping.